### PR TITLE
Added disable limits and home commands

### PR DESCRIPTION
**Pull request by ibaranov-cp**
- Disable limits allows PTU to be driven into hardstops, please use at own risk (damage unlikely, but possible)
- Reset command added, but causes driver crash despite delays and checks. Cause unknown, and the respawn is a bandaid fix.

### CODE REVIEW

**Comment by ibaranov-cp** [comment]
Sorta interested in the amount of "wtf" hair pulling @mikepurvis will do at second comment ;)

File: flir_ptu_driver/include/flir_ptu_driver/driver.h [hunk 1]
```diff
@@ -72,6 +72,9 @@ class PTU
   /** \return true if initialization succeeds. */
   bool initialize();
 
+  /**  \return true if PTU software motion limits are disabled. */
+  bool disable_limits();
```
Comment (mikepurvis) [inline flir_ptu_driver/include/flir_ptu_driver/driver.h]: CamelCase please.

http://wiki.ros.org/CppStyleGuide#Function_.2BAC8_Methods

File: flir_ptu_driver/src/driver.cpp [hunk 2]
```diff
@@ -67,6 +67,14 @@ bool PTU::initialized()
   return !!ser_ && ser_->isOpen() && initialized_;
 }
 
+bool PTU::disable_limits()
+{
+  ser_->write("ld ");  // Disable Limits
+  ser_->read(20);
+  Lim = false;
+  return true;
```
Comment (mikepurvis) [inline flir_ptu_driver/src/driver.cpp]: Is the twenty byte response supposed to be a confirmation or at least an echo? Should we be checking that at all before returning true?

File: flir_ptu_driver/src/driver.cpp [hunk 3]
```diff
@@ -196,11 +205,13 @@ bool PTU::setPosition(char type, float pos, bool block)
...
-  if (count < (type == PTU_TILT ? TMin : PMin) || count > (type == PTU_TILT ? TMax : PMax))
-  {
-    ROS_ERROR("Pan Tilt Value out of Range: %c %f(%d) (%d-%d)\n",
-              type, pos, count, (type == PTU_TILT ? TMin : PMin), (type == PTU_TILT ? TMax : PMax));
-    return false;
+  if (Lim){
+    if (count < (type == PTU_TILT ? TMin : PMin) || count > (type == PTU_TILT ? TMax : PMax))
+    {
+      ROS_ERROR("Pan Tilt Value out of Range: %c %f(%d) (%d-%d)\n",
+                type, pos, count, (type == PTU_TILT ? TMin : PMin), (type == PTU_TILT ? TMax : PMax));
```
Comment (mikepurvis) [inline flir_ptu_driver/src/driver.cpp]: `ROS_ERROR_THROTTLE`

File: flir_ptu_driver/src/driver.cpp [hunk 4]
```diff
@@ -196,11 +205,13 @@ bool PTU::setPosition(char type, float pos, bool block)
   int count = static_cast<int>(pos / getResolution(type));
 
   // Check limits
-  if (count < (type == PTU_TILT ? TMin : PMin) || count > (type == PTU_TILT ? TMax : PMax))
-  {
-    ROS_ERROR("Pan Tilt Value out of Range: %c %f(%d) (%d-%d)\n",
-              type, pos, count, (type == PTU_TILT ? TMin : PMin), (type == PTU_TILT ? TMax : PMax));
-    return false;
+  if (Lim){
```
Comment (mikepurvis) [inline flir_ptu_driver/src/driver.cpp]: Brace on its own line please.

File: flir_ptu_driver/src/node.cpp [hunk 5]
```diff
@@ -33,6 +33,7 @@
 #include <flir_ptu_driver/driver.h>
 #include <ros/ros.h>
 #include <sensor_msgs/JointState.h>
+#include <std_msgs/Bool.h>
 #include <serial/serial.h>
```
Comment (mikepurvis) [inline flir_ptu_driver/src/node.cpp]: Alpha order please.

File: flir_ptu_driver/src/node.cpp [hunk 6]
```diff
@@ -103,7 +106,9 @@ void Node::connect()
   // Query for serial configuration
   std::string port;
   int32_t baud;
+  bool limit;
   ros::param::param<std::string>("~port", port, PTU_DEFAULT_PORT);
+  ros::param::param<bool>("~limits", limit, true);
```
Comment (mikepurvis) [inline flir_ptu_driver/src/node.cpp]: Consider `limits_enabled`?

File: flir_ptu_driver/src/node.cpp [hunk 7]
```diff
@@ -156,6 +167,9 @@ void Node::connect()
   // Subscribers : Only subscribe to the most recent instructions
   m_joint_sub = m_node.subscribe
                 <sensor_msgs::JointState>("cmd", 1, &Node::cmdCallback, this);
+
+  m_reset_sub = m_node.subscribe<std_msgs::Bool>("reset",1, &Node::resetCallback, this);
+
```
Comment (mikepurvis) [inline flir_ptu_driver/src/node.cpp]: No blank line at end of function, space after the comma.

File: flir_ptu_driver/src/driver.cpp (hunk 2 above)
Comment (ibaranov-cp) [reply flir_ptu_driver/src/driver.cpp]: Not really, mostly read to clear buffer, else crash on next publish.
Likelihood of this failing is very low, not done for position commands originally either.

File: flir_ptu_driver/src/driver.cpp (hunk 3 above)
Comment (ibaranov-cp) [reply flir_ptu_driver/src/driver.cpp]: Also the original code, will change.

**Comment by ibaranov-cp** [comment]
@mikepurvis Fixes pushed to master on ibaranov-cp/flir_ptu

**Pull request merged**